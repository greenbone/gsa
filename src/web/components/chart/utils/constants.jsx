/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

export const MENU_PLACEHOLDER_WIDTH = 26;
