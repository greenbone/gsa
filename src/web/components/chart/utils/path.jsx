/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */


class Path {
  constructor() {
    this.paths = [];
    this.closed = false;
  }

  push(command, ...paths) {
    if (!this.closed) {
      this.paths.push(command, ...paths);
    }
    return this;
  }

  close() {
    this.push('Z');
    this.closed = true;
    return this;
  }

  move(x, y) {
    return this.push('M', x, y);
  }

  line(x, y) {
    return this.push('L', x, y);
  }

  arc(
    radiusX,
    radiusY,
    x,
    y,
    {largeArc = 0, sweep = 0, rotationDegree = 0} = {},
  ) {
    return this.push(
      'A',
      radiusX,
      radiusY,
      rotationDegree,
      largeArc,
      sweep,
      x,
      y,
    );
  }

  toString() {
    return this.paths.join(' ');
  }
}

const pathFunc = () => new Path();

export default pathFunc;
// vim: set ts=2 sw=2 tw=80:
