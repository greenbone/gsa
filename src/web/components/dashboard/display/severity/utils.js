/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param {Object} params - The filter values.
 * @param {(string|number)} params.start - The start value of the filter range.
 * @param {(string|number)} params.end - The end value of the filter range.
 * @returns {FilterTerm[]} An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({start, end}) => {
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
