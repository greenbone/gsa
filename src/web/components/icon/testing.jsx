/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */


import {test, expect, testing} from '@gsa/testing';

import {render, fireEvent, act} from 'web/utils/testing';
import Theme from 'web/utils/theme';

import {ICON_SIZE_SMALL_PIXELS} from 'web/hooks/useIconSize';

export const testIcon = Icon => {
  test('should render with default width and height', () => {
    const {element} = render(<Icon />);

    expect(element).toHaveStyleRule('width', ICON_SIZE_SMALL_PIXELS);
    expect(element).toHaveStyleRule('height', ICON_SIZE_SMALL_PIXELS);
  });

  test('should handle click', () => {
    const handler = testing.fn();
    const {element} = render(<Icon onClick={handler} value="1" />);

    fireEvent.click(element);

    expect(handler).toHaveBeenCalledWith('1');
  });
  test('should change appearance when disabled', () => {
    const {element} = render(<Icon disabled={true} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: 'svg path',
    });
  });

  test('should change appearance when inactive', () => {
    const {element} = render(<Icon active={false} />);

    expect(element).toHaveStyleRule('fill', Theme.inputBorderGray, {
      modifier: '& svg path',
    });
  });

  test('should not call clickhandler when disabled', () => {
    const handler = testing.fn();
    const {element} = render(<Icon disabled={true} onClick={handler} />);

    fireEvent.click(element);

    expect(handler).not.toHaveBeenCalled();
  });

  test('should render loading state', async () => {
    let resolve;
    const promise = new Promise((res, rej) => {
      resolve = res;
    });

    const handler = testing.fn().mockReturnValue(promise);

    const {element} = render(
      <Icon title="foo" loadingTitle="bar" onClick={handler} />,
    );

    const svgElement = element.querySelector('svg');

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');

    fireEvent.click(element);

    expect(element).toHaveStyleRule('display', 'none', {
      modifier: '& svg path',
    });
    expect(element).toHaveStyleRule(
      'background',
      'url(/img/loading.gif) center center no-repeat',
      {
        modifier: '& svg',
      },
    );

    expect(element).toHaveAttribute('title', 'bar');
    expect(svgElement).toHaveAttribute('title', 'bar');

    await act(async () => {
      resolve();
    });

    expect(element).toHaveAttribute('title', 'foo');
    expect(svgElement).toHaveAttribute('title', 'foo');
  });
};

// vim: set ts=2 sw=2 tw=80:
