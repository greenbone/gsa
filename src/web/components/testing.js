/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {isDefined} from 'gmp/utils/identity';

import {
  userEvent,
  act,
  fireEvent,
  queryAllByRole,
  queryByRole,
  getByRole,
  getAllByTestId,
  getByTestId,
} from 'web/utils/testing';

export const getElementOrDocument = element =>
  isDefined(element) ? element : document.body;

/**
 * Get all the radio inputs
 */
export const getRadioInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Radio-radio');
};

/**
 * Get all the items of a Select component
 */
export const getSelectItemElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll("[role='option']");
};

/**
 * Get the items that are selectable of a Select component
 */
export const getSelectItemElementsForSelect = async element => {
  element = isDefined(element) ? element : getSelectElement();
  await openSelectElement(element);
  const selectItemsId = element.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector('#' + selectItemsId);
  return getSelectItemElements(itemsContainer);
};

/**
 * Get the input box of a Select component
 */
export const getSelectElement = element => {
  element = getElementOrDocument(element);
  const select = queryByRole(element, 'searchbox');
  if (select) {
    return select;
  }
  return getByRole(element, 'textbox');
};

/**
 * Get all select components
 */
export const getSelectElements = element => {
  element = getElementOrDocument(element);
  return queryAllByRole(element, 'searchbox');
};

/**
 * Open a select element (MultiSelect, Select, etc.)
 */
export const openSelectElement = async select => {
  await act(async () => {
    select = select || getSelectElement();
    await userEvent.click(select);
  });
};

/**
 * Click on an element/item/node
 */
export const clickElement = async element => {
  await act(async () => {
    await userEvent.click(element);
  });
};

/**
 * Get all multi select (root) elements
 */
export const getMultiSelectElements = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-root');
};

/**
 * Get all multi select (root) element
 */
export const getMultiSelectElement = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.mantine-MultiSelect-root');
};

/**
 * Open a MultiSelect component for displaying item selection
 */
export const openMultiSelectElement = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  const input = getSelectElement(element);
  return await clickElement(input);
};

/**
 * Get current selected items of a MultiSelect component
 */
export const getSelectedItems = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-value');
};

/**
 * Get all the selectable items of a MultiSelect component
 */
export const getSelectItemElementsForMultiSelect = async element => {
  element = isDefined(element) ? element : getMultiSelectElement();
  await openMultiSelectElement(element);
  const controlElement = element.querySelector('.mantine-MultiSelect-wrapper');
  const selectItemsId = controlElement.getAttribute('aria-controls');
  const itemsContainer = document.body.querySelector(
    '#' + selectItemsId + '-items',
  );
  return getSelectItemElements(itemsContainer);
};

/**
 * Change the value of an input element like Select or TestField component
 */
export const changeInputValue = (element, value) => {
  fireEvent.change(element, {target: {value}});
};

/**
 * Change the input of a Select component
 */
export const changeSelectInput = (value, input) => {
  if (!isDefined(input)) {
    input = getSelectElement();
  }
  changeInputValue(input, value);
};

/**
 * Query if a dialog is present
 */
export const queryDialog = element => {
  element = getElementOrDocument(element);
  return queryByRole(element, 'dialog');
};

/**
 * Get a dialog
 */
export const getDialog = element => {
  element = getElementOrDocument(element);
  return getByRole(element, 'dialog');
};

/**
 * Get the dialog content
 */
export const getDialogContent = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-body');
};

/**
 * Get the dialog title
 */
export const getDialogTitle = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-title');
};

/**
 * Get the save button of a dialog
 */
export const getDialogSaveButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-save-button');
};

/**
 * Get the close button of a dialog (in the footer)
 */
export const getDialogCloseButton = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-close-button');
};

/**
 * Close a dialog
 */
export const closeDialog = dialog => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  const closeButton = dialog.querySelector('.mantine-CloseButton-root');
  fireEvent.click(closeButton);
};

/**
 * Get the element containing the powerfilter
 */
export const getPowerFilter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('.powerfilter');
};

/**
 * Get text inputs
 */
export const getTextInputs = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-TextInput-input');
};

/**
 * Get the table element
 */
export const getTable = element => {
  element = getElementOrDocument(element);
  return element.querySelector('table');
};

/**
 * Get the table body element
 */
export const getTableBody = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tbody');
};

/**
 * Get the table footer element
 */
export const getTableFooter = element => {
  element = getElementOrDocument(element);
  return element.querySelector('tfoot');
};

/**
 * Get the table header element
 */
export const getTableHeader = element => {
  element = getElementOrDocument(element);
  return element.querySelector('thead');
};

/**
 * Get action items
 */
export const getActionItems = element => {
  element = getElementOrDocument(element);
  return getAllByTestId(element, 'svg-icon');
};

/**
 * Get the bulk action items of a page
 */
export const getBulkActionItems = element => {
  const tableFooter = getTableFooter(element);
  return getActionItems(tableFooter);
};

/**
 * Get the check boxes
 */
export const getCheckBoxes = element => {
  element = getElementOrDocument(element);
  return element.querySelectorAll('.mantine-Checkbox-input');
};
