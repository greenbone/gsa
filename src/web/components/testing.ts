/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {PointerEventsCheckLevel} from '@testing-library/user-event';
import {isDefined} from 'gmp/utils/identity';
import {expect, Mock} from 'vitest';
import {
  userEvent,
  fireEvent,
  queryByRole,
  getByRole,
  getByTestId,
  queryAllByTestId,
  queryAllByRole,
  screen,
} from 'web/utils/Testing';

/**
 * Returns the provided element if it is defined, otherwise returns the document body.
 *
 * @param element - The element to check.
 * @returns The provided element if defined, otherwise the document body.
 */
export const getElementOrReturnDocument = (element?: HTMLElement) =>
  isDefined(element) ? element : document.body;

/**
 * Get all the radio inputs within the specified element.
 *
 * @param element - The optional parent element or document to search within.
 * @returns A NodeList of radio input elements.
 */
export const getRadioInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-Radio-radio');
};

/**
 * Get all the items of a Select component
 *
 * @param element - The optional root element or document to search within.
 * @returns A NodeList of select item elements
 */
export const getSelectItemElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll("[role='option']");
};

/**
 * Retrieves the selectable item elements for a given select element.
 *
 * @param element - The optional select element to retrieve items for. If not provided, a default select element will be used.
 * @returns A promise that resolves to a NodeList of the selectable item elements.
 */
export const getSelectItemElementsForSelect = async (
  element?: HTMLSelectElement,
) => {
  element = isDefined(element) ? element : getSelectElement();
  await openSelectElement(element);
  const selectItemsId = element.getAttribute('aria-controls');
  const itemsContainer =
    document.body.querySelector<HTMLElement>('#' + selectItemsId) || undefined;
  return getSelectItemElements(itemsContainer);
};

/**
 * Get the input box of a Select component
 *
 * This function first attempts to find an element with the test ID 'form-select'.
 * If no such element is found, it falls back to finding an element with the role 'textbox'.
 *
 * @param element - The optional element to search within.
 * @returns The found select element
 */
export const getSelectElement = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  const select = getByTestId(element, 'form-select');
  if (select) {
    return select as HTMLSelectElement;
  }
  return getByRole(element, 'textbox') as HTMLSelectElement;
};

/**
 * Get all select components
 *
 * @param element - The root element to query within. If not provided, the document will be used.
 * @returns A list of select elements
 */
export const queryAllSelectElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryAllByTestId(element, 'form-select');
};

/**
 * Open a select element (MultiSelect, Select, etc.)
 *
 * @param select - An optional select element to open. If not provided, it defaults to the result of getSelectElement().
 * @returns A promise that resolves when the select element has been opened.
 */
export const openSelectElement = async (select?: HTMLSelectElement) => {
  select = select || getSelectElement();
  await clickElement(select);
};

/**
 * Clicks on the given element/item/node using userEvent.
 *
 * @param element - The element to be clicked.
 * @returns A promise that resolves when the click action is complete.
 */
export const clickElement = async (element: Element) => {
  await userEvent.click(element, {
    pointerEventsCheck: PointerEventsCheckLevel.Never,
  });
};

/**
 * Get all multi select (root) elements
 *
 * @param element - An optional element to search within. If not provided, the document will be used.
 * @returns A NodeList of multi select elements.
 */
export const getMultiSelectElements = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-MultiSelect-inputField');
};

/**
 * Retrieves the multi select (root) element from the given element.
 *
 * @returns The found multi select element.
 */
export const getMultiSelectElement = () => {
  return screen.getByRole('textbox');
};

/**
 * Open a MultiSelect component for displaying item selection
 *
 * @param element - The multi-select element to open. If not defined, it will use a default multi-select element.
 * @returns A promise that resolves when the input element has been clicked.
 */
export const openMultiSelectElement = async (element?: HTMLElement) => {
  element = isDefined(element) ? element : getMultiSelectElement();
  const input = getSelectElement(element);
  return await clickElement(input);
};

/**
 * Get current selected items of a MultiSelect component
 *
 * @param element - The document object to query.
 * @returns A NodeList of selected items.
 */
export const getSelectedItems = (element: ParentNode) => {
  return element.querySelectorAll<HTMLElement>('.mantine-MultiSelect-pill');
};

/**
 * Get all the selectable items of a MultiSelect component from screen.
 *
 * @returns An array of select item elements.
 */
export const getSelectItemElementsForMultiSelect = () => {
  return screen.getAllByRole('option');
};

/**
 * Changes the value of an input element like Select or TestField component and triggers the blur event.
 *
 * @param element - The input element whose value is to be changed.
 * @param value - The new value to be set for the input element.
 */
export const changeInputValue = (
  element: Document | Element | Window | Node,
  value: string,
) => {
  fireEvent.change(element, {target: {value}});
  fireEvent.blur(element);
};

/**
 * Changes the value of a select input element.
 *
 * @param value - The new value to set for the select input.
 * @param input - The select input element to change. If not provided, a default select element will be used.
 */
export const changeSelectInput = (value: string, input?: HTMLSelectElement) => {
  if (!isDefined(input)) {
    input = getSelectElement();
  }
  changeInputValue(input, value);
};

/**
 * Queries all dialog elements within the specified element or document.
 *
 * @param element - The element or document to query within. If not provided, the document will be used.
 * @returns A list of dialogs.
 */
export const queryDialogs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryAllByRole(element, 'dialog');
};

/**
 * Queries if a dialog is present
 *
 * @param element - The element to query within. If not provided, the document will be used.
 * @returns The dialog element if found, otherwise null.
 */
export const queryDialog = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return queryByRole(element, 'dialog');
};

/**
 * Retrieves the dialog
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The dialog element found within the specified element or document.
 */
export const getDialog = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return getByRole(element, 'dialog');
};

/**
 * Retrieves the content of a dialog.
 *
 * @param dialog - The dialog element. If not defined, a default dialog is retrieved.
 * @returns The content element of the dialog, or null if not found.
 */
export const queryDialogContent = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-body');
};

/**
 * Retrieves the title element from a dialog.
 *
 * @param dialog - The dialog element. If not provided, a default dialog is retrieved using the getDialog function.
 * @returns The title element of the dialog, or null if not found.
 */
export const queryDialogTitle = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return dialog.querySelector('.mantine-Modal-title');
};

/**
 * Retrieves the save button of a dialog (in the footer)
 *
 * @param dialog - The dialog element. If not provided, the default dialog will be used.
 * @returns The save button element within the dialog.
 */
export const getDialogSaveButton = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-save-button');
};

/**
 * Retrieves the close button of a dialog (in the footer).
 *
 * @param dialog - The dialog element. If not provided, the default dialog will be used.
 * @returns The close button element within the dialog.
 */
export const getDialogCloseButton = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  return getByTestId(dialog, 'dialog-close-button');
};

/**
 * Closes a dialog by clicking its close (X) button.
 *
 * @param dialog - The dialog element to close. If not provided, it defaults to the result of getDialog().
 */
export const closeDialog = (dialog?: HTMLElement) => {
  dialog = isDefined(dialog) ? dialog : getDialog();
  const closeButton = dialog.querySelector(
    '.mantine-CloseButton-root',
  ) as HTMLElement;
  fireEvent.click(closeButton);
};

/**
 * Get the element containing the powerfilter
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The powerfilter element, or null if not found.
 */
export const queryPowerFilter = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('.powerfilter');
};

/**
 * Get text inputs
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of text input elements
 */
export const queryTextInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-TextInput-input');
};

/**
 * Queries the table element
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The first table element found within the given element, or null if no table is found.
 */
export const queryTable = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('table');
};

/**
 * Queries the table body element
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The <tbody> element if found, otherwise null.
 */
export const queryTableBody = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('tbody');
};

/**
 * Queries the table footer element.
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The `<tfoot>` element if found, otherwise `null`.
 */
export const queryTableFooter = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('tfoot');
};

/**
 * Queries the table header element.
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns The <thead> element of the table, or null if not found.
 */
export const queryTableHeader = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelector('thead');
};

/**
 * Queries and returns all checkbox input elements
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of all checkbox input elements found within the specified element.
 */
export const queryCheckBoxes = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-Checkbox-input');
};

/**
 * Queries and returns all file input elements
 *
 * @param element - The element to search within. If not provided, the document will be used.
 * @returns A NodeList of file input elements
 */
export const queryFileInputs = (element?: HTMLElement) => {
  element = getElementOrReturnDocument(element);
  return element.querySelectorAll('.mantine-FileInput-input');
};

export const testBulkTrashcanDialog = (
  _screen: unknown,
  dialogAction: Mock,
) => {
  const dialog = screen.getByRole('dialog');
  expect(dialog).toBeVisible();

  const moveToTrashcanButton = screen.getByText('Move to Trashcan');
  fireEvent.click(moveToTrashcanButton);

  expect(dialogAction).toHaveBeenCalled();
};

export const testBulkDeleteDialog = (_screen: unknown, dialogAction: Mock) => {
  const dialog = screen.getByTestId('confirmation-dialog');
  expect(dialog).toBeVisible();

  const title = screen.getByText('Confirm Deletion');

  expect(title).toBeVisible();

  const deleteButton = screen.getByText('Delete');
  fireEvent.click(deleteButton);

  expect(dialogAction).toHaveBeenCalled();
};
