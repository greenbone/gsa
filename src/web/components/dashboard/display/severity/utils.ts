/* SPDX-FileCopyrightText: 2025 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import FilterTerm from 'gmp/models/filter/filterterm';
import {isDefined} from 'gmp/utils/identity';
import {LOG_VALUE} from 'web/utils/severity';

interface FilterValueToFilterTermsParams {
  start?: number;
  end?: number;
}

type FilterValueToFilterTermsReturnType = [FilterTerm?, FilterTerm?];

/**
 * Converts filter values to filter terms based on severity range.
 *
 * @param params - The filter values.
 * @param params.start - The start value of the filter range.
 * @param params.end - The end value of the filter range.
 * @returns An array of FilterTerm objects representing the filter terms.
 */
export const filterValueToFilterTerms = ({
  start,
  end,
}: FilterValueToFilterTermsParams): FilterValueToFilterTermsReturnType => {
  if (!isDefined(start)) {
    return [];
  }
  if (isDefined(end)) {
    const startTerm = FilterTerm.fromString(`severity>${start}`);
    const endTerm = FilterTerm.fromString(`severity<${end}`);
    return [startTerm, endTerm];
  }
  if (start > LOG_VALUE) {
    return [FilterTerm.fromString(`severity>${start}`)];
  }
  return [FilterTerm.fromString(`severity=${start}`)];
};
