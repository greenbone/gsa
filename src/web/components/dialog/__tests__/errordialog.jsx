/* SPDX-FileCopyrightText: 2024 Greenbone AG
 *
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {describe, test, expect, testing} from '@gsa/testing';

import {KeyCode} from 'gmp/utils/event';

import {render, fireEvent} from 'web/utils/testing';

import ErrorDialog from '../errordialog';

describe('ErrorDialog component tests', () => {
  test('should render ErrorDialog with text and title', () => {
    const handleClose = testing.fn();

    const {baseElement, getByTestId} = render(
      <ErrorDialog text="foo" title="bar" onClose={handleClose} />,
    );

    expect(baseElement).toMatchSnapshot();
    const contentElement = getByTestId('errordialog-content');
    const titleElement = getByTestId('dialog-title-bar');
    expect(contentElement).toHaveTextContent('foo');
    expect(titleElement).toHaveTextContent('bar');
  });

  test('should close ErrorDialog with close button', () => {
    const handleClose = testing.fn();

    const {getByTestId} = render(
      <ErrorDialog title="bar" onClose={handleClose} />,
    );

    const closeButton = getByTestId('dialog-title-close-button');
    fireEvent.click(closeButton);
    expect(handleClose).toHaveBeenCalled();
  });

  test('should close ErrorDialog with resume button', () => {
    const handleClose = testing.fn();

    const {baseElement} = render(
      <ErrorDialog title="bar" onClose={handleClose} />,
    );

    const buttons = baseElement.querySelectorAll('button');
    fireEvent.click(buttons[0]);
    expect(handleClose).toHaveBeenCalled();
  });

  test('should close ErrorDialog on escape key', () => {
    const handleClose = testing.fn();

    const {getByRole} = render(
      <ErrorDialog title="bar" onClose={handleClose} />,
    );

    fireEvent.keyDown(getByRole('dialog'), {
      key: 'Escape',
      keyCode: KeyCode.ESC,
    });

    expect(handleClose).toHaveBeenCalled();
  });
});
