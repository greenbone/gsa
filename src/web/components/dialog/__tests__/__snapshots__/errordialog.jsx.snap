// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorDialog component tests > should render ErrorDialog with text and title 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 10% auto;
  border: 0;
  outline: 0;
  width: 400px;
  height: autopx;
}

.c0 {
  position: fixed;
  font-family: Trebuchet MS,Tahoma,Verdana,Arial,sans-serif;
  font-size: 1.1em;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  margin: 0;
  background: rgba(102, 102, 102, 0.5);
  z-index: 600;
  -webkit-transition: opacity 1s ease-in;
  transition: opacity 1s ease-in;
  width: 100%;
  height: 100%;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: inherit;
  padding: 0;
  background: #fff;
  box-shadow: 5px 5px 10px #7F7F7F;
  border-radius: 3px;
  border: 1px solid #7F7F7F;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: bold;
  font-size: 12px;
  font-family: Verdana,sans-serif;
  color: #074320;
  cursor: pointer;
  border-radius: 2px;
  padding: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c5 :hover {
  border: 1px solid #074320;
}

.c6 {
  height: 24px;
  width: 24px;
  line-height: 24px;
}

.c6 * {
  height: inherit;
  width: inherit;
}

.c12 {
  display: inline-block;
  padding: 0 15px;
  color: #4C4C4C;
  text-align: center;
  vertical-align: middle;
  font-size: 11px;
  font-weight: bold;
  line-height: 30px;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  background-color: #fff;
  border-radius: 2px;
  border: 1px solid #bfbfbf;
  cursor: pointer;
  overflow: visible;
  z-index: 1;
}

.c12:focus,
.c12:hover {
  border: 1px solid #4C4C4C;
}

.c12:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background: #11ab51;
  font-weight: bold;
  color: #fff;
}

.c12[disabled] {
  cursor: not-allowed;
  opacity: 0.65;
  box-shadow: none;
}

.c12 img {
  height: 32px;
  width: 32px;
  margin-top: 5px 10px 5px -10px;
  vertical-align: middle;
}

.c12:link {
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #4C4C4C;
}

.c13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c14 {
  border: 1px solid #7F7F7F;
  color: #fff;
  background: #11ab51;
}

.c14 :hover {
  color: #074320;
  background: #A1DDBA;
}

.c11 {
  border-width: 1px 0 0 0;
  border-style: solid;
  border-color: #e5e5e5;
  margin-top: 15px;
  padding: 10px 20px 10px 20px;
}

.c4 {
  padding: 5px 5px 5px 10px;
  margin-bottom: 15px;
  border-radius: 2px 2px 0 0;
  border-bottom: 1px solid #7F7F7F;
  color: #fff;
  font-weight: bold;
  background: #11ab51;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: -webkit-grab;
  cursor: grab;
}

.c9 {
  overflow: auto;
  padding: 0 15px;
  width: 100%;
  height: 100%;
}

.c8 {
  overflow: hidden;
  height: 100%;
}

<body>
  <div
    id="portals"
  >
    <div>
      <div
        class="c0"
      >
        <div
          class="c1"
          height="autopx"
          role="dialog"
          tabindex="1"
          width="400px"
        >
          <div
            class="c2"
          >
            <div
              class="c3 c4"
              data-testid="dialog-title-bar"
            >
              <span>
                bar
              </span>
              <div
                class="c5 c6"
                data-testid="dialog-title-close-button"
                title="Close"
              >
                Ã—
              </div>
            </div>
            <div
              class="c7 c8"
            >
              <div
                class="c9"
                data-testid="errordialog-content"
              >
                foo
              </div>
            </div>
            <div
              class="c10 c11"
            >
              <button
                class="c12 c13 c14"
                title="OK"
              >
                OK
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
